### PR TITLE
[C/PyTorch/Jax] Add support for more bias shapes

### PR DESCRIPTION
**Pull request by cyanguwa**
This PR
- adds support for `[b,1,s,s]`,`[b,h,s,s]`, `[1,1,s,s]` bias shapes, when dBias is not required. This is applicable to inference (`bias.requires_grad = False`), when bias is a workaround for an arbitrary mask (`True/False -> 0/-inf`), or when ALiBi slopes tensor is in `[b,h]` shape.
- makes changes to the `F16_arbitrary_seqlen` backend of the cuDNN fused attention only.
- makes changes to C, PyTorch and Jax attention implementations.

### CODE REVIEW

**Comment by denera** [comment]
@cyanguwa I merged my TE/JAX updates into this PR. Still need to update the unit tests to include the new bias shapes.

**Comment by cyanguwa** [comment]
Pytorch pipepline with CUDA 12.3: 13014488. Fixing A100 errors now.

**Comment by cyanguwa** [comment]
Pipeline 13043757 for PyTorch and Paddle.

**Comment by denera** [comment]
With the latest commit, *all* FP16 tests in TE/JAX CI are now passing with `neg_inf = -2^15`

The BF16 failures look like this:
- Any input with `max_seqlen > 512` + any mask
- Inputs with `max_seqlen <= 512` and `max_seqlen_q != max_seqlen_kv` + `NO_MASK`

It's not clear to me yet if these BF16 failures are due to a bug in the pure-JAX/Flax reference function or the TE/JAX fused attn custom op.

File: tests/jax/test_fused_attn.py [hunk 1]
```diff
@@ -174,27 +189,62 @@ def _check_configs(self):
...
+        if self.bias_shape != BiasShape.BIAS_1HSS:
+            if self.attn_bias_type != AttnBiasType.POST_SCALE_BIAS:
+                pytest.skip("B1SS, BHSS and 11SS bias shapes require POST_SCALE_BIAS.")
+            elif self.attn_mask_type not in [AttnMaskType.NO_MASK, AttnMaskType.CAUSAL_MASK]:
+                pytest.skip("B1SS, BHSS and 11SS bias shapes are only supported for "
+                            "AttnMaskType.NO_MASK and AttnMaskType.CAUSAL_MASK.")
+            elif self.backend != NVTE_Fused_Attn_Backend.NVTE_F16_arbitrary_seqlen:
+                pytest.skip("B1SS, BHSS and 11SS bias shapes are only supported for "
+                            "the F16_arbitrary_seqlen backend.")
+
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:208]: The logic here should be integrated into [nvte_get_fused_attn_backend](https://github.com/NVIDIA/TransformerEngine/blob/main/transformer_engine/common/include/transformer_engine/fused_attn.h#L146) and return no_backend for the unsupported combinations.

File: tests/jax/test_fused_attn.py [hunk 2]
```diff
@@ -217,7 +267,7 @@ def test_forward(self):
         """
         Test forward without JIT
         """
-        self._setup_inputs()
+        self._setup_inputs(True)
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:270]: Not sure why we need this

File: tests/jax/test_fused_attn.py [hunk 3]
```diff
@@ -236,19 +286,13 @@ def test_forward(self):
...
-        np.testing.assert_allclose(primitive_invalid, jnp.zeros_like(primitive_invalid))
+        assert_allclose(primitive_valid, reference_valid, dtype=self.dtype)
+        assert_allclose(primitive_invalid, jnp.zeros_like(primitive_invalid), dtype=self.dtype)
 
     def test_backward(self):
         """
         Test value_and_grad with JIT, which includes both forward and backward
         """
-        if not self.is_training:
-            pytest.skip("Backward doesn't support inference")
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:251]: Could you insight me why `is_training` and `dropout_prob` can be dropped here?

File: tests/jax/test_fused_attn.py [hunk 4]
```diff
@@ -338,42 +387,53 @@ def check_dqkv(primitive, reference, valid_len):
...
+    pytest.param(jnp.bfloat16, id="BF16"),
+    pytest.param(jnp.float16, id="FP16")
+])
+@pytest.mark.parametrize('b, s_q, s_kv, h_q, h_kv, d',[
+    pytest.param(32,  128,  128, 16, 16, 64, id='SELF_512'),
+    pytest.param( 4, 2048, 2048, 12, 12, 64, id='SELF_ARB'),
+    pytest.param(32,  512,  128, 16, 16, 64, id='CROSS_512'),
+    pytest.param( 4, 2048, 1024, 12, 12, 64, id='CROSS_ARB'),
+    pytest.param(32,  128,  128, 12,  6, 64, id='GQA_512'),
+    pytest.param( 4, 2048, 2048, 12,  6, 64, id='GQA_ARB')
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:404]: I preferred the original naming so that we can regonize the failed tests at the first time instead of checking the source file especially for the CI.

File: tests/jax/test_fused_attn.py [hunk 5]
```diff
@@ -338,42 +387,53 @@ def check_dqkv(primitive, reference, valid_len):
...
+    ])
     @staticmethod
-    def test_forward(b, s_q, s_kv, h_q, h_kv, d, attn_bias_type, attn_mask_type, dropout_prob,
-                     dtype, is_training, qkv_layout):
+    def test_forward(b, s_q, s_kv, h_q, h_kv, d, attn_bias_type, attn_mask_type,
+                     dropout_prob, dtype, is_training, qkv_layout, bias_shape):
         """
         Test forward with parameterized configs
         """
+        pytest.skipif(is_training and dropout_prob > 0.,
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:425]: oh, I just see you move the skipif to here, but this is not the original intend. The original tests still run the custom call for the sanity check and only skips the value checks. However, moving here then we will lose the sanity check for the dropout.

File: tests/jax/test_fused_attn.py [hunk 6]
```diff
@@ -149,6 +156,13 @@ def customcall_fused_dpa(query, key, value, bias, q_token, kv_token, dropout_rng
                               **kwargs).astype(query.dtype)
 
 
+class BiasShape(Enum):
+    BIAS_1HSS = '1HSS'
+    BIAS_B1SS = 'B1SS'
+    BIAS_BHSS = 'BHSS'
+    BIAS_11SS = '11SS'
```
Comment (zlsh80826) [inline tests/jax/test_fused_attn.py:163]: Nitpick, `BiashShape.BIAS_1HSS` looks duplicated for the "bias" term.

File: transformer_engine/jax/cpp_extensions.py [hunk 7]
```diff
@@ -1904,12 +1904,22 @@ def abstract(qkv_aval, bias_aval, seqlen_or_cu_seqlen_aval, seed_aval, *, attn_b
         rng_state_shape = (seed_aval.shape[0], checker.rng_state_size)
         rng_state_aval = seed_aval.update(shape=rng_state_shape, dtype=checker.rng_state_dtype)
 
+        # NOTE: We can't rely on checking bias type here because we want to support arbitrary bias
+        #       shapes the bias tensor is being used as a workaround for an arbitrary mask
+        #       (True/False -> 0/-Inf) with bias type set to NO_BIAS.
```
Comment (zlsh80826) [inline transformer_engine/jax/cpp_extensions.py:1909]: Could you help me understand in which case the bias_type is `NO_BIAS` but the bias_aval.shape is not zero?

Comment (denera) [reply transformer_engine/jax/cpp_extensions.py:1909]: This is an incorrect comment. The arbitrary mask workaround is supported via `AttnBiasType.POST_SCALE_BIAS` and the code is implemented for it but I neglected to update the comment here.

File: tests/jax/test_fused_attn.py (hunk 5 above)
Comment (denera) [reply tests/jax/test_fused_attn.py:425]: Ah, that makes sense. I missed this and thought that we were basically just doing a lot of wasteful compute here without testing anything. I'll move it back to get the sanity check.

File: tests/jax/test_fused_attn.py (hunk 4 above)
Comment (denera) [reply tests/jax/test_fused_attn.py:404]: Yes, this is a temporary change here just to make the printouts a bit more concise/readable for me while debugging the CI failures. I will change them back to the original IDs before finalizing the PR.

File: tests/jax/test_fused_attn.py (hunk 2 above)
Comment (denera) [reply tests/jax/test_fused_attn.py:270]: Yes, leftover from debugging. I'll be doing a lot of code cleanup once the CI failures are fully fixed and will re-request a review at that time.

File: tests/jax/test_fused_attn.py (hunk 1 above)
Comment (denera) [reply tests/jax/test_fused_attn.py:208]: I tentatively agree with this. If we intend to support this arbitrary mask workaround as a proper feature, we should have an `AttnBiasLayout` as part of the backend check.

@cyanguwa perhaps we can revisit this once I clean up the CI failures?

File: tests/jax/test_fused_attn.py (hunk 3 above)
Comment (denera) [reply tests/jax/test_fused_attn.py]: The pytest skip condition here means `test_backward()` never executes when `is_training = False`, so there's no point for `@pytest.mark.parametrize('is_training',...)` to be a general parameter for the entire test class. I moved it to `test_forward()` only instead to produce a cleaner pytest output with fewer skips.

I saw your point about the dropout though and I will be bringing that back to maintain the sanity check.

**Review by zlsh80826 — APPROVED** [review]
LGTM
